tests/test-firefox/test.firefox.js

```diff
@@ -480,29 +481,53 @@ describe('firefox', () => {
         });
     });
 
-    it('throws an error when the port is occupied', () => {
-      // TODO: add a retry for occupied ports.
-      // https://github.com/mozilla/web-ext/issues/283
-      const client = fake(RemoteFirefox.prototype);
-      const connectToFirefox = sinon.spy(() => Promise.resolve(client));
-      return findRemotePort({connectToFirefox})
-        .then(makeSureItFails())
-        .catch(onlyInstancesOf(WebExtError, (error) => {
-          assert.match(error.message, /Cannot listen on port/);
-          assert.equal(client.disconnect.called, true);
+    it('returns a port on first try', () => {
+      const connectToFirefox = sinon.spy(() => new Promise(
+        (resolve, reject) => {
+          reject(
+            new TCPConnectError('first call connection fails - port is free')
+          );
         }));
+      return findRemotePort({connectToFirefox})
+      .then((port) => {
+        assert.equal(connectToFirefox.callCount, 1);
+        assert.isNumber(port);
+      });
     });
 
-    it('re-throws unexpected connection errors', () => {
-      const connectToFirefox = sinon.spy(
-        () => Promise.reject(new Error('not a connection error')));
+    it('cancel discovery after too many fails', () => {
+      const client = fake(RemoteFirefox.prototype);
+      const connectToFirefox = sinon.spy(() => new Promise(
+        (resolve) => {
+          resolve(client);
+        }));
       return findRemotePort({connectToFirefox})
-        .then(makeSureItFails())
-        .catch((error) => {
-          assert.match(error.message, /not a connection error/);
+      .catch((err) => {
+        assert.equal(err, 'WebExtError: Too many retries on port search');
+      });
+    });
+
+    it('retries port discovery after first failure', () => {
+      const client = new RemoteFirefox(fakeFirefoxClient());
+      const connectToFirefox = sinon.spy((port) => {
+        return new Promise((resolve, reject) => {
+          if (port===6006) {
```

Comment:
relying on the port number will make this test fragile because the starting port number may be configured differently in the future. A better approach would be to use a counter variable and simply reject the *first* call to `connectToFirefox` but not the second call. For example:

````javascript
let callCount = 1
const connectToFirefox = sinon.spy((port) => new Promise((resolve, reject) => {
  if (callCount === 1) {
    // Pretend like the first attempt connects to a port in use:
    resolve(client);
  } else {
    // Pretend like the second attempt finds an open port:
    reject(new TCPConnectError('port is free'));
  }
  callCount ++;
});
````